_posts/2020-10-23-deconvolution.md

```diff
@@ -0,0 +1,446 @@
+images.  Specifically, we looked at `deconvolution`, a common method to
+_deblur_ images.  Now, a year later, we return to these experiments with a
+better understanding of how Dask and CuPy can interact, enhanced serialization
+methods, and support from the open-source community.
+
+Image Analysis Redux
+--------------------
+
+Previously we used the [Richardson Lucy
+(RL)](https://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution)
+deconvolution algorithm from ITK and
+[Scikit-Image](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/deconvolution.py#L329).
+We left off at theorizing how GPUs could potentially help accelerate these
+workflows.  Starting with Scikit-Image's implementation, we naively tried
+replacing `scipy.signal.convolve` calls with `cupyx.scipy.ndimage.convolve`,
+and while performance improved, it did not improve _significantly_ -- that is,
+we did not get the 100X speed we were looking for.
+
+As it often turns out in mathematics a problem that can be inefficient to solve
+in one representation can often be made more efficent by transforming the data
+beforehand. In this new representation we can solve the same problem
+(convolution in this case) more easily before transforming the result back into
+a more familiar representation. When it comes to convolution, the
+transformation we apply is called [Fast-Fourirer Transform
+(FFT)](https://en.wikipedia.org/wiki/Fast_Fourier_transform). Once this is
+applied we are able to convolve data using a simple multiplication.
+
+As it turns out this FFT transformation is extremely fast on both CPUs and
+GPUs. Similarly the algorithm we can write with FFTs are accelerated.  Despite
+the added step of doing FFTs, the cost of transformation + the cost of the
+algorithm is still lower than performing the original algorithm in real space.
+We (and others before us) found this was the case for Richardson Lucy (on both
+CPUs and GPUs) and performance continued increasing when we parallelized with
+Dask over multiple GPUs.
+
+Help from Open-Source
+---------------------
+
+An FFT RL equivalent has been around for some time and the good folks at the
+[Solar Dynamics Observatory](https://sdo.gsfc.nasa.gov/mission/instruments.php)
+built and shared a NumPy/CuPy implementation as part the [Atmospheric Imaging
+Assembly](https://aiapy.readthedocs.io/en/v0.2.0/_modules/aiapy/psf/deconvolve.html)
+Python package (aiapy).  We slightly modified their implementation to handle 3D
+as well as 2D [Point Spread
+Functions](https://en.wikipedia.org/wiki/Point_spread_function) and to take
+advantage of
+[NEP-18](https://numpy.org/neps/nep-0018-array-function-protocol.html) for
+convenient dispatching of NumPy and CuPy to NumPy and CuPy functions:
+
+
+```python
+def deconvolve(img, psf=None, iterations=20):
+    """
+    """
+    # 2D vs 3D psfs
+    if len(psf.shape) == 2:
+        psf = psf
+    else:
+        psf = psf[0]
```

Comment:
Style nit and suggested diff:

``` diff
-     # 2D vs 3D psfs
-     if len(psf.shape) == 2:
-         psf = psf
-     else:
-         psf = psf[0]
+ psf = psf if psf.ndim == 2 else psf[0]  # 2D vs 3D PSFs
```

Reply:
I find it slightly easier in the blog to see this spelled out as opposed to one line. That said, there may be other changes here anyways.

Reply:
This code has since been dropped